File: packages/http-client-java/generator/http-client-generator-test/src/test/java/client/naming/NamingTests.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package client.naming;

import client.naming.model.models.ClientModel;
import client.naming.model.models.JavaModel;
import client.naming.property.models.ClientNameAndJsonEncodedNameModel;
import client.naming.property.models.ClientNameModel;
import client.naming.property.models.LanguageClientNameModel;
import client.naming.unionenum.models.ClientExtensibleEnum;
import client.naming.unionenum.models.ExtensibleEnum;
import org.junit.jupiter.api.Test;

public class NamingTests {

    private final NamingClient client = new NamingClientBuilder().buildClient();

    // client name should be "ClientModel", currently a bug in TCGC
    private final ClientModelClient modelClient = new NamingClientBuilder().buildClientModelClient();

    private final UnionEnumClient enumClient = new NamingClientBuilder().buildUnionEnumClient();

    @Test
    public void testNaming() {
        // operation
        client.clientName();

        // parameter
        client.parameter("true");

        // property
        client.client(new ClientNameModel(true));
        client.language(new LanguageClientNameModel(true));
        client.compatibleWithEncodedName(new ClientNameAndJsonEncodedNameModel(true));

        // header
        client.request("true");
        client.response();  // no class about response header at present

        // model
        modelClient.client(new ClientModel(true));
        modelClient.language(new JavaModel(true));

        // enum
        enumClient.unionEnumName(ClientExtensibleEnum.ENUM_VALUE1);
        enumClient.unionEnumMemberName(ExtensibleEnum.CLIENT_ENUM_VALUE1);
    }
}

File: packages/http-client-csharp/generator/TestProjects/CadlRanch.Tests/Http/Client/Naming/NamingTests.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Threading.Tasks;
using Client.Naming;
using Client.Naming.Models;
using NUnit.Framework;
using ClientModel = Client.Naming.Models.ClientModel;

namespace TestProjects.CadlRanch.Tests.Http.Client.Naming
{
    public class ClientNamingTests : CadlRanchTestBase
    {
        [CadlRanchTest]
        public Task Client() => Test(async (host) =>
        {
            var response = await new NamingClient(host, null).ClientAsync(new ClientNameModel(true));
            Assert.AreEqual(204, response.GetRawResponse().Status);

            Assert.NotNull(typeof(ClientNameModel).GetProperty("ClientName"));
            Assert.IsNull(typeof(ClientNameModel).GetProperty("DefaultName"));
        });

        [CadlRanchTest]
        public Task Language() => Test(async (host) =>
        {
            var response = await new NamingClient(host, null).LanguageAsync(new LanguageClientNameModel(true));
            Assert.AreEqual(204, response.GetRawResponse().Status);

            Assert.NotNull(typeof(LanguageClientNameModel).GetProperty("CSName"));
            Assert.IsNull(typeof(LanguageClientNameModel).GetProperty("DefaultName"));
        });

        [CadlRanchTest]
        public Task CompatibleWithEncodedName() => Test(async (host) =>
        {
            var response = await new NamingClient(host, null).CompatibleWithEncodedNameAsync(new ClientNameAndJsonEncodedNameModel(true));
            Assert.AreEqual(204, response.GetRawResponse().Status);

            Assert.NotNull(typeof(ClientNameModel).GetProperty("ClientName"));
            Assert.IsNull(typeof(ClientNameModel).GetProperty("DefaultName"));
        });

        [CadlRanchTest]
        public Task Operation() => Test(async (host) =>
        {
            var response = await new NamingClient(host, null).ClientNameAsync();
            Assert.AreEqual(204, response.GetRawResponse().Status);
        });

        [CadlRanchTest]
        public Task Parameter() => Test(async (host) =>
        {
            var response = await new NamingClient(host, null).ParameterAsync(clientName: "true");
            Assert.AreEqual(204, response.GetRawResponse().Status);
        });

        [CadlRanchTest]
        public Task Request() => Test(async (host) =>
        {
            var response = await new NamingClient(host, null).RequestAsync(clientName: "true");
            Assert.AreEqual(204, response.GetRawResponse().Status);
        });

        [CadlRanchTest]
        public Task Response() => Test(async (host) =>
        {
            var response = await new NamingClient(host, null).ResponseAsync();
            Assert.IsTrue(response.GetRawResponse().Headers.TryGetValue("default-name", out _));
            foreach (var header in response.GetRawResponse().Headers)
            {
                var key = header.Key;
                if (key == "default-name")
                {
                    var value = header.Value;
                    Assert.AreEqual("true", value);
                }
            }
        });

        [CadlRanchTest]
        public Task ModelClient() => Test(async (host) =>
        {
            var response = await new NamingClient(host, null).GetClientModelClient().ClientAsync(new ClientModel(true));
            Assert.AreEqual(204, response.GetRawResponse().Status);
        });

        [CadlRanchTest]
        public Task ModelLanguage() => Test(async (host) =>
        {
            var response = await new NamingClient(host, null).GetClientModelClient().LanguageAsync(new CSModel(true));
            Assert.AreEqual(204, response.GetRawResponse().Status);
        });

        [CadlRanchTest]
        public Task UnionEnumName() => Test(async (host) =>
        {
            var response = await new NamingClient(host, null).GetUnionEnumClient().UnionEnumNameAsync(ClientExtensibleEnum.EnumValue1);
            Assert.AreEqual(204, response.GetRawResponse().Status);
        });

        [CadlRanchTest]
        public Task UnionEnumMemberName() => Test(async (host) =>
        {
            var response = await new NamingClient(host, null).GetUnionEnumClient().UnionEnumMemberNameAsync(ExtensibleEnum.ClientEnumValue1);
            Assert.AreEqual(204, response.GetRawResponse().Status);
        });
    }
}
